zulipterminal/ui_tools/boxes.py

```diff
@@ -476,14 +501,24 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
                     # Prepend org url to convert it to an absolute link
                     link = urljoin(self.model.server_url, link)
 
-                if link == text:
-                    # If the link and text are same
-                    # usually the case when user just pastes
-                    # a link then just display the link
-                    markup.append(('msg_link', text))
+                text = text if text else link
+
+                # Detect duplicate links to save screen real estate.
+                if link not in self.message_links:
+                    self.message_links[link] = (text, self.link_index)
+                    self.link_index += 1
                 else:
-                    markup.append(
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                    saved_text, saved_link_index = self.message_links[link]
```

Comment:
The else bit is to retain `text` if the link is already present in `message_links`.